### DISCUSSION TITLE
Remix Tailwind Vercel Edge problems

### DISCUSSION THREAD

**Discussion by JasonColeyNZ** [remix-run/remix#8204] [General]
Hi there,

I am trying to deploy a remix app to the vercel edge, I use tailwind extensively in the app and this uses tw-merge (Tailwind Merge).
But this falls over on vercel edge. I get...

Error: The Edge Function "404" is referencing unsupported modules:
--
10:44:01.535 | - build/build-edge-eyJydW50aW1lIjoiZWRnZSJ9.js: tailwind-merge

Is there any known workaround to use tw-merge on vercel edge. I find it almost incomprehensible that no-one else has hit this?

**Comment by JasonColeyNZ** [comment]
OK managed to get it working, I'm using esm, and I don't understand exactly what the issue is, but if I copy the tailwind-merge src folder to my utils and use this instead of the npm package then it all works fine, so I'm up and running, but would be really nice to understand why this is required, maybe an issue with tailwind-merge and esm?

**Comment by cxreiff** [comment]
I ran into this issue as well- I ran into an identical issue with the "bcryptjs" package, and had to copy that as well. It seems there are some fairly frustrating foot-guns somewhere in the locus of ESM, Remix, the edge runtime, and the way that certain dependencies are configured, but I haven't been able to figure out exactly what is going on. Both this package and "bcryptjs" are very widely used, it seems to me that more and more folks will run into this as more adopt native ESM and edge environments.

**Comment by cxreiff** [comment]
Oh! While the following solution didn't work for "bcrypt.js", it appears to work fine for this package: if you add `serverDependenciesToBundle: ["tailwind-merge"],` to your remix.config.js, it appears to correctly make the module available during the deployment.

**Reply by dananjaya6005** [reply]
if i have vite.config.ts where should i add this ?

**Reply by rmacfie** [reply]
I believe this is the equivalent for `vite.config.ts`: 

```ts
export default defineConfig({
  plugins: [
    remix(),
    tsconfigPaths(),
  ],
  build: {
    rollupOptions: {
      external: ['tailwind-merge'],
    },
  },
});
```

**Reply by sahithyandev** [reply]
@rmacfie It's not working :(

**Reply by DuncanLHS** [reply]
I was suffering the same issue so simply avoided having any UI running on edge for the time being.

This worked for a while but the error re-appeared as soon as I started to add UI to my root.tsx. It seems that everything being imported in the root file is being bundled to run on edge. Despite not being told to!

The only place I'm choosing to use edge is on some server functions.
(P.S. @rmacfie your solution is not working for me)

**Comment by ovtheo** [comment]
> if i have vite.config.ts where should i add this ?

@dananjaya6005 Here is the code that worked for me. Inspired by https://github.com/remix-run/remix/discussions/8204#discussioncomment-7774661 and https://github.com/remix-run/remix/issues/7865#issuecomment-1791236651. Hope this helps!
```
export default defineConfig({
    plugins: [
        remix(),
        tsconfigPaths(),
    ],
    ssr: {
        noExternal: ["tailwind-merge"]
    }
});
```
